include/continuable/support/asio.hpp

```diff
@@ -0,0 +1,84 @@
+/*
+
+                        /~` _  _ _|_. _     _ |_ | _
+                        \_,(_)| | | || ||_|(_||_)|(/_
+
+                    https://github.com/Naios/continuable
+                                   v4.0.0
+
+  Copyright(c) 2015 - 2019 Denis Blank <denis.blank at outlook dot com>
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy
+  of this software and associated documentation files(the "Software"), to deal
+  in the Software without restriction, including without limitation the rights
+  to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+  copies of the Software, and to permit persons to whom the Software is
+  furnished to do so, subject to the following conditions :
+
+  The above copyright notice and this permission notice shall be included in
+  all copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  SOFTWARE.
+**/
+
+#ifndef CONTINUABLE_SUPPORT_ASIO_HPP_INCLUDED
+#define CONTINUABLE_SUPPORT_ASIO_HPP_INCLUDED
+
+#include <continuable/continuable-base.hpp>
+#include <continuable/detail/support/asio.hpp>
+#include <continuable/detail/utility/traits.hpp>
+
+CTI_DETAIL_ASIO_NAMESPACE_BEGIN
+
+// Class used to specify an asynchronous operation should return a continuable.
+struct use_cti_t {};
+
+// Special value for instance of `use_cti_t`.
+constexpr use_cti_t use_cti{};
+
+template <typename Signature>
+class async_result<use_cti_t, Signature> {
+public:
+#if defined(CTI_DETAIL_ASIO_HAS_EXPLICIT_RET_TYPE_INTEGRATION)
+  using return_type = typename cti::detail::asio::initiate_make_continuable<
+      Signature>::erased_return_type;
+#endif
+
+  template <typename Initiation, typename... Args>
+  static
+#if defined(CTI_DETAIL_ASIO_HAS_EXPLICIT_RET_TYPE_INTEGRATION)
+      return_type
+#else
+      auto
+#endif
+      initiate(Initiation initiation, use_cti_t, Args... args) {
```

Comment:
Can't we just leave the return type auto? The explicit `return_type` type only needs to be returned in case asio is checking: `std::is_same<decltype(initiate(...)), return_type>::value`. Otherwise the return type will just converted to `return_type` itself.

Reply:
Yes you're totally right. ASIO is not doing such checks, it just uses `ASIO_INITFN_AUTO_RESULT_TYPE(token, signature)` to be `auto` if ret type deduction is available, else `typename async_result<token, signature>::return_type`.